File: baekjoon/2252.py
from collections import deque
from sys import stdin


def main():
    node_count, edge_count = (int(s) for s in input().split())
    incomings: list[int] = [0 for _ in range(node_count)]
    nexts: list[list[int]] = [[] for _ in range(node_count)]
    for _ in range(edge_count):
        node_a, node_b = (int(s) for s in stdin.readline().split())
        node_a -= 1
        node_b -= 1
        nexts[node_a].append(node_b)
        incomings[node_b] += 1
    sorted_nodes = sort_nodes(incomings, nexts)
    print(" ".join(str(i + 1) for i in sorted_nodes))


def sort_nodes(incomings: list[int], nexts: list[list[int]]) -> list[int]:
    sorted_nodes: list[int] = []
    cursors = deque[int]()

    for node, incoming in enumerate(incomings):
        if incoming == 0:
            cursors.append(node)

    while cursors:
        current_node = cursors.popleft()
        sorted_nodes.append(current_node)
        for next_node in nexts[current_node]:
            incomings[next_node] -= 1
            if incomings[next_node] == 0:
                cursors.append(next_node)

    return sorted_nodes


main()

File: baekjoon/2252.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Node {
  int incomings;
  vector<int> nexts;
};

auto sort_nodes(vector<Node>& nodes) -> vector<int> {
  int node_count = static_cast<int>(nodes.size());
  vector<int> sorted_nodes;
  sorted_nodes.reserve(node_count);

  queue<int> bfs_queue;
  for (int i = 0; i < node_count; i++) {
    if (nodes[i].incomings == 0) {
      bfs_queue.push(i);
    }
  }

  while (!bfs_queue.empty()) {
    int idx_a = bfs_queue.front();
    bfs_queue.pop();
    sorted_nodes.push_back(idx_a);
    for (int idx_b : nodes[idx_a].nexts) {
      Node& next_node = nodes[idx_b];
      next_node.incomings -= 1;
      if (next_node.incomings == 0) {
        bfs_queue.push(idx_b);
      }
    }
  }

  return sorted_nodes;
}

auto main() -> int {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  cout.tie(nullptr);

  int node_count = 0;
  int edge_count = 0;
  cin >> node_count >> edge_count;

  vector<Node> nodes;
  nodes.reserve(node_count);
  for (int i = 0; i < node_count; i++) {
    nodes.push_back({0, {}});
  }

  for (int i = 0; i < edge_count; i++) {
    int node_a = 0;
    int node_b = 0;
    cin >> node_a >> node_b;
    node_a -= 1;
    node_b -= 1;
    nodes[node_a].nexts.push_back(node_b);
    nodes[node_b].incomings += 1;
  }

  vector<int> sorted_nodes = sort_nodes(nodes);
  for (int node : sorted_nodes) {
    cout << (node + 1) << ' ';
  }
}